src/config/app/domain.rs

```diff
@@ -0,0 +1,110 @@
+use std::error::Error;
+use std::fmt;
+
+static RESERVED_PACKAGE_NAMES: [&'static str; 2] = ["kotlin", "java"];
+
+#[derive(Debug)]
+pub enum DomainError {
+    Empty,
+    NotAsciiAlphanumeric { bad_chars: Vec<char> },
+    StartsWithDigit { label: String },
+    ReservedPackageName { package_name: String },
+    StartsOrEndsWithADot,
+    EmptyLabel,
+}
+
+impl Error for DomainError {}
+
+impl fmt::Display for DomainError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Empty => write!(f, "Domain can't be empty"),
+            Self::NotAsciiAlphanumeric { bad_chars } => write!(
+                f,
+                "\"{}\" are not valid ASCII alphanumeric characters",
+                bad_chars.into_iter().collect::<String>()
+            ),
+            Self::ReservedPackageName { package_name } => {
+                write!(f, "\"{}\" is reserved and cannot be used", package_name)
+            }
+            Self::StartsWithDigit { label } => write!(
+                f,
+                "\"{}\" label starts with a digit, which is invalid",
```

Comment:
This might not be compelling to users, since anyone who owns/uses a domain that starts with a number is going to disagree with us. It'd be fantastic to make it clear that it's invalid for Java package names, especially since then the code self-documents why we have this constraint.

This argument also applies to forbidding hyphens, non-ASCII, and reserved names... do you think it'd be more elegant to keep the old `publicsuffix` domain syntax validation as a first pass with the previous generic "invalid domain syntax" error, and then have the relevant parts of the new validation as a second pass that is specifically for disallowing things that don't play nice with Java/Kotlin? That should make the nature of the constraints more obvious, and probably involve a bit less code. Tell me if I'm wrong, though!

Reply:
Normally I'd agree that using `publicsuffix` first and they having our own, more strict validation would be beneficial, but there are a few things I don't like about domain validation in publicsuffix:
```rust
pub fn has_valid_syntax(domain: &str) -> bool {
        // we are explicitly checking for this here before calling `domain_to_ascii`
        // because `domain_to_ascii` strips of leading dots so we won't be able to
        // check for this later
        if domain.starts_with('.') { return false; }
        // let's convert the domain to ascii early on so we can validate
        // internationalised domain names as well
        let domain = match Self::to_ascii(domain) {
            Ok(domain) => { domain }
            Err(_) => { return false; }
        };
        let mut labels: Vec<&str> = domain.split('.').collect();
        // strip of the first dot from a domain to support fully qualified domain names
        if domain.ends_with(".") { labels.pop(); }
        // a domain must not have more than 127 labels
        if labels.len() > 127 { return false; }
        labels.reverse();
        for (i, label) in labels.iter().enumerate() {
            // the tld must not be a number
            if i == 0 && label.parse::<f64>().is_ok() { return false; }
            // any label must only contain allowed characters
            if !LABEL.is_match(label) { return false; }
        }
        true
    }
```
First of all, the check is just true/false and there's no way to return an information to the user about which part or which characters are wrong. Also, using this function wouldn't reduce much code because, besides our obious java & kotlin specific constraints we'd need to check ascii alphanumeric again (as `has_valid_syntax` converts utf-8 to ascii first using [IDN](https://en.wikipedia.org/wiki/Internationalized_domain_name) and it's obviously something we can't use).
We'd also need to iterate over all the labels and validate them one more time which is just simply wasteful imo.

What do you think about this?

(As for domains starting with numbers, I changed the information to more appropriate one :slightly_smiling_face:)

Reply:
Gotcha! In that case, I totally agree. Thanks for being so thorough!

Reply:
No problem, thanks for the feedback! I'll resolve this conversation then